File: day2/solution.js
const { readFileSync } = require('fs');
const { join } = require('path');

const solveFirstPuzzle = (input) => {
  let solution = 0;

  input.replace(/,/g, ';').split('\n').map(l => l.split(': ')[1]).forEach((line, id) => {
    const colors = line.split('; ');

    let ok = true;

    colors.forEach((color) => {
      const [amount, name] = color.split(' ');
      
      if ((name === 'blue' && +amount > 14) || (name === 'green' && +amount > 13) || (name === 'red' && +amount > 12)) {
        ok = false;
      }
    });

    if (ok) {
      solution += (id+1);
    }
  });

  return solution
}

const solveSecondPuzzle = (input) => {
  let solution = 0;

  input.replace(/,/g, ';').split('\n').map(l => l.split(': ')[1]).forEach(line => {
    const colors = line.split('; ');

    let blue = 0;
    let green = 0;
    let red = 0;
    
    colors.forEach((color) => {
      const [amount, name] = color.split(' ');
      
      if (name === 'blue' && blue < Number(amount)) {
        blue = Number(amount);
      } else if (name === 'green' && green < Number(amount)) {
        green = Number(amount);
      } else if (name === 'red' && red < Number(amount)) {
        red = Number(amount);
      }
    });

    solution += blue * green * red;
  });

  return solution
}

console.log(solveFirstPuzzle(readFileSync(join(__dirname, 'input.txt'), 'utf8')));
console.log(solveSecondPuzzle(readFileSync(join(__dirname, 'input.txt'), 'utf8')));

File: day2/solution.cpp
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

void firstProblem(std::vector<std::string> &lines)
{
    int solution = 0;

    for (int i = 0; i < lines.size(); i++)
    {
        std::string line = lines[i];

        for (int k = 0; k < line.length(); k++)
        {
            if (line[k] == ',')
            {
                line.replace(k, 1, ";");
            }
        }

        std::string startDelimiter = ": ";
        std::string ballsDelimiter = "; ";

        line.erase(0, line.find(startDelimiter) + startDelimiter.length());

        std::vector<std::string> balls;
        size_t pos = 0;

        while ((pos = line.find(ballsDelimiter)) != std::string::npos)
        {
            balls.push_back(line.substr(0, pos));
            line.erase(0, pos + ballsDelimiter.length());
        }

        balls.push_back(line);

        bool ok = true;

        for (int j = 0; j < balls.size(); j++)
        {
            std::string ballDelimiter = " ";

            std::string amount = balls[j].substr(0, balls[j].find(ballDelimiter));
            balls[j].erase(0, balls[j].find(ballDelimiter) + ballDelimiter.length());

            std::string color = balls[j];

            if (color == "red" && std::stoi(amount) > 12)
            {
                ok = false;
            }
            if (color == "green" && std::stoi(amount) > 13)
            {
                ok = false;
            }
            if (color == "blue" && std::stoi(amount) > 14)
            {
                ok = false;
            }
        }

        if (ok)
        {
            solution += (i + 1);
        }
    }

    std::cout << solution << std::endl;
}

void secondProblem(std::vector<std::string> &lines)
{
    int solution = 0;

    for (int i = 0; i < lines.size(); i++)
    {
        std::string line = lines[i];

        for (int k = 0; k < line.length(); k++)
        {
            if (line[k] == ',')
            {
                line.replace(k, 1, ";");
            }
        }

        std::string startDelimiter = ": ";
        std::string ballsDelimiter = "; ";

        line.erase(0, line.find(startDelimiter) + startDelimiter.length());

        std::vector<std::string> balls;
        size_t pos = 0;

        while ((pos = line.find(ballsDelimiter)) != std::string::npos)
        {
            balls.push_back(line.substr(0, pos));
            line.erase(0, pos + ballsDelimiter.length());
        }

        balls.push_back(line);

        int red = 0;
        int green = 0;
        int blue = 0;

        for (int j = 0; j < balls.size(); j++)
        {
            std::string ballDelimiter = " ";

            std::string amount = balls[j].substr(0, balls[j].find(ballDelimiter));
            balls[j].erase(0, balls[j].find(ballDelimiter) + ballDelimiter.length());

            std::string color = balls[j];

            if (std::stoi(amount) > red && color == "red")
            {
                red = std::stoi(amount);
            }
            if (std::stoi(amount) > green && color == "green")
            {
                green = std::stoi(amount);
            }
            if (std::stoi(amount) > blue && color == "blue")
            {
                blue = std::stoi(amount);
            }
        }

        solution += (red * green * blue);
    }

    std::cout << solution << std::endl;
}

int main()
{
    std::ifstream myFile;

    myFile.open("input.txt");

    if (!myFile.is_open())
    {
        std::cout << "Unable to open file" << std::endl;
        return 1;
    }

    std::string line;
    std::vector<std::string> lines;

    while (std::getline(myFile, line))
    {
        if (line.empty())
        {
            break;
        }

        lines.push_back(line);
    }

    firstProblem(lines);
    secondProblem(lines);

    return 0;
}
